File: web/src/types/direct-message.ts
import type { User } from "./auth";

export type MessageType = "TEXT" | "IMAGE" | "PDF" | "DOCUMENT";

export interface DirectMessageAttachment {
  key: string;
  url: string;
  originalName: string;
  mimeType: string;
  size: number;
}

export interface DirectMessageReaction {
  emoji: string;
  count?: number;
  users: string[];
  category?: "STANDARD" | "CAMPUS_CUSTOM";
}

export interface DirectMessageReply {
  _id: string;
  content: string;
  deleted: boolean;
  senderId?: Pick<User, "_id" | "fullName">;
}

export interface DirectMessage {
  _id: string;
  conversationId: string;
  senderId: Pick<User, "_id" | "fullName" | "rollNumber" | "profilePicture">;
  content: string;
  messageType: MessageType;
  attachments: DirectMessageAttachment[];
  replyTo?: DirectMessageReply;
  reactions?: DirectMessageReaction[];
  deletedFor?: string[];
  isDeletedForEveryone?: boolean;
  deletedBy?: string | { _id: string; fullName: string };
  deletedAt?: string;
  edited: boolean;
  editedAt?: string;
  read: boolean;
  readAt?: string;
  deleted: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface ConversationLastMessage {
  content: string;
  senderId: string;
  createdAt: string;
}

export interface Conversation {
  _id: string;
  participants: Pick<User, "_id" | "fullName" | "rollNumber" | "profilePicture" | "department">[];
  lastMessage: ConversationLastMessage | null;
  lastMessageAt: string | null;
  isLocked?: boolean;
  lockedBy?: string | { _id: string; fullName: string };
  lockedReason?: string;
  lockedAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface PaginatedConversations {
  items: Conversation[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

export interface PaginatedDirectMessages {
  items: DirectMessage[];
  total: number;
  page: number;
  limit: number;
  pages: number;
  order: "latest" | "oldest";
}

File: backend-java/src/main/java/com/studyconnect/backend/entity/DirectMessage.java
package com.studyconnect.backend.entity;

import com.studyconnect.backend.entity.enums.MessageType;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.FieldType;
import org.springframework.data.mongodb.core.mapping.MongoId;

@Getter
@Setter
@NoArgsConstructor
@Document(collection = "directmessages")
public class DirectMessage {

    @Id
    @MongoId(FieldType.OBJECT_ID)
    private String id;

    @Indexed
    private String conversationId;

    @Indexed
    private String senderId;

    private String content = "";
    private MessageType messageType = MessageType.TEXT;
    private List<DirectMessageAttachment> attachments = new ArrayList<>();
    @Indexed
    private String replyTo;
    private boolean edited = false;
    private Instant editedAt;
    private boolean read = false;
    private Instant readAt;
    private boolean deleted = false;
    private Instant deletedAt;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    @Getter
    @Setter
    @NoArgsConstructor
    public static class DirectMessageAttachment {
        private String key;
        private String url;
        private String originalName;
        private String mimeType;
        private long size;
    }
}
